<a href="https://colab.research.google.com/github/harishmuh/Python-for-Data-Science-Analysis/blob/main/Pandas_III_Exercise.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Exercise - Pandas III**
---

**Importing libraries**

In [1]:
# Importing libraries
import numpy as np
import pandas as pd

You can find the required datasets here:

* audi.csv : [Download here](https://github.com/harishmuh/Python-for-Data-Science-Analysis/blob/main/datasets/audi.csv) | [Url link](https://raw.githubusercontent.com/harishmuh/Python-for-Data-Science-Analysis/refs/heads/main/datasets/audi.csv)
* hyundai.csv: [Download here](https://github.com/harishmuh/Python-for-Data-Science-Analysis/blob/main/datasets/hyundai.csv) | [Url link]()
* toyota.csv: [Download here](https://github.com/harishmuh/Python-for-Data-Science-Analysis/blob/main/datasets/toyota.csv) | [Url link](https://raw.githubusercontent.com/harishmuh/Python-for-Data-Science-Analysis/refs/heads/main/datasets/toyota.csv)

**Loading dataset**

In [2]:
audi_url = 'https://raw.githubusercontent.com/harishmuh/Python-for-Data-Science-Analysis/refs/heads/main/datasets/audi.csv'
df_audi = pd.read_csv(audi_url)
df_audi.head()

,model,year,price,transmission,mileage,fuelType,tax,mpg,engineSize
0,A1,2017,12500,Manual,15735,Petrol,150,55.4,1.4
1,A6,2016,16500,Automatic,36203,Diesel,20,64.2,2.0
2,A1,2016,11000,Manual,29946,Petrol,30,55.4,1.4
3,A4,2017,16800,Automatic,25952,Diesel,145,67.3,2.0
4,A3,2019,17300,Manual,1998,Petrol,145,49.6,1.0


In [3]:
hyundai_url = 'https://raw.githubusercontent.com/harishmuh/Python-for-Data-Science-Analysis/refs/heads/main/datasets/hyundai.csv'
df_hyundai = pd.read_csv(hyundai_url)
df_hyundai.head()

,model,year,price,transmission,mileage,fuelType,tax(£),mpg,engineSize
0,I20,2017,7999,Manual,17307,Petrol,145,58.9,1.2
1,Tucson,2016,14499,Automatic,25233,Diesel,235,43.5,2.0
2,Tucson,2016,11399,Manual,37877,Diesel,30,61.7,1.7
3,I10,2016,6499,Manual,23789,Petrol,20,60.1,1.0
4,IX35,2015,10199,Manual,33177,Diesel,160,51.4,2.0


In [4]:
toyota_url = 'https://raw.githubusercontent.com/harishmuh/Python-for-Data-Science-Analysis/refs/heads/main/datasets/toyota.csv'
df_toyota = pd.read_csv(toyota_url)
df_toyota.head()

,model,year,price,transmission,mileage,fuelType,tax,mpg,engineSize
0,GT86,2016,16000,Manual,24089,Petrol,265,36.2,2.0
1,GT86,2017,15995,Manual,18615,Petrol,145,36.2,2.0
2,GT86,2015,13998,Manual,27469,Petrol,265,36.2,2.0
3,GT86,2017,18998,Manual,14736,Petrol,150,36.2,2.0
4,GT86,2017,17498,Manual,36284,Petrol,145,36.2,2.0


## **No.1**

- If you notice, the column name 'tax(£)' in the Hyundai dataset is different from the other datasets.
- Change the column name to 'tax' to match the other datasets.

In [5]:
df_hyundai = df_hyundai.rename(columns={'tax(£)' : 'tax'})
df_hyundai.head()

,model,year,price,transmission,mileage,fuelType,tax,mpg,engineSize
0,I20,2017,7999,Manual,17307,Petrol,145,58.9,1.2
1,Tucson,2016,14499,Automatic,25233,Diesel,235,43.5,2.0
2,Tucson,2016,11399,Manual,37877,Diesel,30,61.7,1.7
3,I10,2016,6499,Manual,23789,Petrol,20,60.1,1.0
4,IX35,2015,10199,Manual,33177,Diesel,160,51.4,2.0


## **No.2**

- Create a new column named 'brand' and enter the car brand name in the leftmost column.
- For example, for Audi data frames, enter 'Audi' in the 'brand' column!

In [6]:
# # method 1
# df_toyota['brand']='Toyota' # appears at the very back

In [7]:
# Method 2
df_toyota.insert(0, 'brand', 'Toyota') # (index, column_name, content) --> no need to reassign
df_audi.insert(0, 'brand', 'Audi')
df_hyundai.insert(0, 'brand', 'Hyundai')

# error, when rerun/column name already exists

## **No.3**

- Combine the three data frames into a single data frame named df_car. Ensure that no indexes are repeated.

In [8]:
df_car = pd.concat([df_audi, df_hyundai, df_toyota], axis=0).reset_index(drop=True)
df_car.head()

,brand,model,year,price,transmission,mileage,fuelType,tax,mpg,engineSize
0,Audi,A1,2017,12500,Manual,15735,Petrol,150,55.4,1.4
1,Audi,A6,2016,16500,Automatic,36203,Diesel,20,64.2,2.0
2,Audi,A1,2016,11000,Manual,29946,Petrol,30,55.4,1.4
3,Audi,A4,2017,16800,Automatic,25952,Diesel,145,67.3,2.0
4,Audi,A3,2019,17300,Manual,1998,Petrol,145,49.6,1.0


## **No.4**

- Display dataframe information regarding column names, the number of data points, and their data types.
- Provide an explanation of the dataframe information!

In [9]:
df_car.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22266 entries, 0 to 22265
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   brand         22266 non-null  object 
 1   model         22266 non-null  object 
 2   year          22266 non-null  int64  
 3   price         22266 non-null  int64  
 4   transmission  22266 non-null  object 
 5   mileage       22266 non-null  int64  
 6   fuelType      22266 non-null  object 
 7   tax           22266 non-null  int64  
 8   mpg           22266 non-null  float64
 9   engineSize    22266 non-null  float64
dtypes: float64(2), int64(4), object(4)
memory usage: 1.7+ MB


- The DataFrame consists of 22266 rows and 10 columns.
- 4 columns are integers and 2 columns are floats.
- 4 columns are strings (objects).
- There are no missing values.

## **No.5**

- Show if there are any missing values ​​in the dataframe!
- Show if there are any duplicate data in the table!
- If there are, handle the missing values ​​and duplicate data!

In [10]:
# check for missing values
df_car.isna().sum().to_frame('missing values')

# no missing values

,missing values
brand,0
model,0
year,0
price,0
transmission,0
mileage,0
fuelType,0
tax,0
mpg,0
engineSize,0


In [11]:
# count the number of duplicates
df_car.duplicated().sum()

# there are 228 duplicates

np.int64(228)

In [12]:
# data dimensions before removing duplicates
df_car.shape

(22266, 10)

In [13]:
# remove duplicates
df_car = df_car.drop_duplicates().reset_index(drop=True)

In [14]:
# data dimensions after removing duplicates
df_car.shape

(22038, 10)

## **No.6**

- Display statistics for numeric columns!
- Provide an explanation of the statistical information obtained!

In [15]:
df_car.describe()

,year,price,mileage,tax,mpg,engineSize
count,22038.000000,22038.000000,22038.000000,22038.000000,22038.000000,22038.000000
mean,2016.982303,17522.086033,23613.609946,115.353934,55.208531,1.690453
std,2.134123,10587.356894,21063.202824,69.024931,14.851751,0.566645
min,1997.000000,850.000000,1.000000,0.000000,1.100000,0.000000
25%,2016.000000,9999.000000,7446.250000,30.000000,45.600000,1.400000
50%,2017.000000,14999.000000,18539.500000,145.000000,55.400000,1.600000
75%,2019.000000,21990.000000,33186.750000,145.000000,64.200000,2.000000
max,2020.000000,145000.000000,323000.000000,580.000000,256.800000,6.300000


In general, we can see that:
* Based on **Year** data, the car's production year ranges from 1997 to 2020.
* **Price** ranges from 850 to 145,000, with an average of 17,522.
* 75% of cars sold have a mileage of under 33,186 miles, but there are also cars with a mileage of up to 323,000 miles. There are even cars that can be considered new with a mileage of 1 mile.
* **Tax** issued ranges from 0 to 580, with the most frequent value being 145.
* The most fuel-efficient car has an mpg of 1.1, while the most fuel-efficient car has an mpg of 256.8.
* There are cars with an engine size of 0.

## **No.7**

- Display statistics for categorical columns!
- Provide an explanation of the statistical information obtained!

In [17]:
df_car.describe(include='object')

,brand,model,transmission,fuelType
count,22038,22038,22038,22038
unique,3,60,4,4
top,Audi,Yaris,Manual,Petrol
freq,10565,2117,11656,11896


* Audi is the most frequently featured brand compared to Toyota and Hyundai.
* The Yaris is the most frequently featured car model, appearing 2,117 times, or approximately 10% of total cars sold.
* Approximately 50% of cars sold have a manual transmission.
* Approximately 50% of cars sold are gasoline-powered.

## **No.8**

Create a dataframe containing three columns:
- Column Name: Column name
- Number of Uniques: Number of unique values
- Unique Values: Unique data/values

In [18]:
# Method 1 (one by one)
number_of_unique = df_car.nunique()
number_of_unique

,0
brand,3
model,60
year,24
price,5113
transmission,4
mileage,15386
fuelType,4
tax,40
mpg,132
engineSize,26


In [19]:
uniques = df_car.apply(lambda x: x.unique())
uniques

,0
brand,"[Audi, Hyundai, Toyota]"
model,"[ A1, A6, A4, A3, Q3, Q5, A5, S4, Q2, ..."
year,"[2017, 2016, 2019, 2015, 2014, 2018, 2013, 202..."
price,"[12500, 16500, 11000, 16800, 17300, 13900, 132..."
transmission,"[Manual, Automatic, Semi-Auto, Other]"
mileage,"[15735, 36203, 29946, 25952, 1998, 32260, 7678..."
fuelType,"[Petrol, Diesel, Hybrid, Other]"
tax,"[150, 20, 30, 145, 125, 200, 0, 205, 160, 235,..."
mpg,"[55.4, 64.2, 67.3, 49.6, 58.9, 61.4, 70.6, 60...."
engineSize,"[1.4, 2.0, 1.0, 3.0, 1.6, 1.8, 1.5, 4.0, 2.5, ..."


In [20]:
pd.DataFrame({
    'Column Name' : number_of_unique.index,
    'Number of Unique' : number_of_unique.values,
    'Unique Values' : uniques
})

,Column Name,Number of Unique,Unique Values
brand,brand,3,"[Audi, Hyundai, Toyota]"
model,model,60,"[ A1, A6, A4, A3, Q3, Q5, A5, S4, Q2, ..."
year,year,24,"[2017, 2016, 2019, 2015, 2014, 2018, 2013, 202..."
price,price,5113,"[12500, 16500, 11000, 16800, 17300, 13900, 132..."
transmission,transmission,4,"[Manual, Automatic, Semi-Auto, Other]"
mileage,mileage,15386,"[15735, 36203, 29946, 25952, 1998, 32260, 7678..."
fuelType,fuelType,4,"[Petrol, Diesel, Hybrid, Other]"
tax,tax,40,"[150, 20, 30, 145, 125, 200, 0, 205, 160, 235,..."
mpg,mpg,132,"[55.4, 64.2, 67.3, 49.6, 58.9, 61.4, 70.6, 60...."
engineSize,engineSize,26,"[1.4, 2.0, 1.0, 3.0, 1.6, 1.8, 1.5, 4.0, 2.5, ..."


In [21]:
# Method 2 (For Loop)
list_unik = []

for kolom in df_car:
    list_unik.append([kolom, df_car[kolom].nunique(), df_car[kolom].unique()])

pd.DataFrame(list_unik, columns=['Column Name', 'Number of Unique', 'Unique Values'])

,Column Name,Number of Unique,Unique Values
0,brand,3,"[Audi, Hyundai, Toyota]"
1,model,60,"[ A1, A6, A4, A3, Q3, Q5, A5, S4, Q2, ..."
2,year,24,"[2017, 2016, 2019, 2015, 2014, 2018, 2013, 202..."
3,price,5113,"[12500, 16500, 11000, 16800, 17300, 13900, 132..."
4,transmission,4,"[Manual, Automatic, Semi-Auto, Other]"
5,mileage,15386,"[15735, 36203, 29946, 25952, 1998, 32260, 7678..."
6,fuelType,4,"[Petrol, Diesel, Hybrid, Other]"
7,tax,40,"[150, 20, 30, 145, 125, 200, 0, 205, 160, 235,..."
8,mpg,132,"[55.4, 64.2, 67.3, 49.6, 58.9, 61.4, 70.6, 60...."
9,engineSize,26,"[1.4, 2.0, 1.0, 3.0, 1.6, 1.8, 1.5, 4.0, 2.5, ..."


## **No.9**

- In the 'model' column, some of the data contains unnecessary whitespace at the beginning or end.
- Remove this whitespace from all data in the 'model' column!

In [22]:
df_car['model'] = df_car['model'].str.strip()
df_car.head()

,brand,model,year,price,transmission,mileage,fuelType,tax,mpg,engineSize
0,Audi,A1,2017,12500,Manual,15735,Petrol,150,55.4,1.4
1,Audi,A6,2016,16500,Automatic,36203,Diesel,20,64.2,2.0
2,Audi,A1,2016,11000,Manual,29946,Petrol,30,55.4,1.4
3,Audi,A4,2017,16800,Automatic,25952,Diesel,145,67.3,2.0
4,Audi,A3,2019,17300,Manual,1998,Petrol,145,49.6,1.0


## **No.10**

Save df_car in .csv format with the file name 'all_cars.csv'!

In [23]:
# save df_car to csv with name 'all_cars.csv'
df_car.to_csv('all_cars.csv')

## **No.11**

- From df_car, display the average tax based on make and transmission. Then, sort by highest tax.

In [24]:
df_average_tax = df_car.groupby(['brand', 'transmission']).mean(numeric_only=True)[['tax']]
df_average_tax.sort_values('tax', ascending=False)

,,tax
brand,transmission,
Hyundai,Automatic,151.610942
Audi,Automatic,149.453358
Hyundai,Semi-Auto,148.767606
Audi,Semi-Auto,143.522982
Toyota,Semi-Auto,124.015748
Hyundai,Manual,111.152002
Toyota,Manual,104.016610
Audi,Manual,96.735001
Toyota,Automatic,78.217654


In [25]:
df_car.groupby(['brand', 'transmission']).mean(numeric_only=True)[['tax']].sort_values('tax', ascending=False)

,,tax
brand,transmission,
Hyundai,Automatic,151.610942
Audi,Automatic,149.453358
Hyundai,Semi-Auto,148.767606
Audi,Semi-Auto,143.522982
Toyota,Semi-Auto,124.015748
Hyundai,Manual,111.152002
Toyota,Manual,104.016610
Audi,Manual,96.735001
Toyota,Automatic,78.217654


## **No.12**

- You are asked to find a Toyota Camry with a price below the average Camry price and a mileage below the average Camry mileage.
- Show cars that match these criteria!

In [28]:
# dataframe containing Camry cars
Camry_car = df_car[df_car['model']=='Camry']
Camry_car

,brand,model,year,price,transmission,mileage,fuelType,tax,mpg,engineSize
22012,Toyota,Camry,2019,29990,Automatic,4600,Hybrid,135,52.3,2.5
22013,Toyota,Camry,2019,25990,Automatic,6000,Hybrid,135,52.3,2.5
22014,Toyota,Camry,2019,24990,Automatic,5145,Hybrid,135,52.3,2.5
22015,Toyota,Camry,2020,26490,Automatic,10000,Hybrid,135,52.3,2.5
22016,Toyota,Camry,2019,27545,Automatic,5059,Hybrid,135,52.3,2.5
22017,Toyota,Camry,2019,28495,Automatic,4998,Hybrid,140,52.3,2.5
22018,Toyota,Camry,2019,26491,Automatic,4416,Hybrid,135,52.3,2.5
22019,Toyota,Camry,2019,27550,Automatic,4189,Hybrid,135,52.3,2.5
22020,Toyota,Camry,2019,25490,Automatic,6340,Hybrid,135,52.3,2.5
22021,Toyota,Camry,2019,26990,Automatic,3000,Hybrid,135,52.3,2.5


In [29]:
# Calculate the average price of a Camry car
avg_price_camry = Camry_car['price'].mean()
avg_price_camry

np.float64(26910.090909090908)

In [30]:
# calculate the average mileage of a Camry car
avg_mileage_camry = Camry_car['mileage'].mean()
avg_mileage_camry

np.float64(5861.727272727273)

In [31]:
Camry_car[(Camry_car['price'] < avg_price_camry) & (Camry_car['mileage'] < avg_mileage_camry)]

,brand,model,year,price,transmission,mileage,fuelType,tax,mpg,engineSize
22014,Toyota,Camry,2019,24990,Automatic,5145,Hybrid,135,52.3,2.5
22018,Toyota,Camry,2019,26491,Automatic,4416,Hybrid,135,52.3,2.5


## **No.13**

Show the most expensive cars!

In [32]:
df_car[df_car['price'] == df_car['price'].max()]

,brand,model,year,price,transmission,mileage,fuelType,tax,mpg,engineSize
4743,Audi,R8,2020,145000,Semi-Auto,2000,Petrol,145,21.1,5.2


## **No.14**

In which year did the most Toyota cars appear?

In [33]:
# Method 1
df_car[df_car['brand']=='Toyota']['year'].mode()[0]

np.int64(2017)

In [34]:
# Method 2
# Step 1: create a table containing only Toyota cars
Toyota_car = df_car[df_car['brand']=='Toyota']
Toyota_car

,brand,model,year,price,transmission,mileage,fuelType,tax,mpg,engineSize
15339,Toyota,GT86,2016,16000,Manual,24089,Petrol,265,36.2,2.0
15340,Toyota,GT86,2017,15995,Manual,18615,Petrol,145,36.2,2.0
15341,Toyota,GT86,2015,13998,Manual,27469,Petrol,265,36.2,2.0
15342,Toyota,GT86,2017,18998,Manual,14736,Petrol,150,36.2,2.0
15343,Toyota,GT86,2017,17498,Manual,36284,Petrol,145,36.2,2.0
...,...,...,...,...,...,...,...,...,...,...
22033,Toyota,IQ,2011,5500,Automatic,30000,Petrol,20,58.9,1.0
22034,Toyota,Urban Cruiser,2011,4985,Manual,36154,Petrol,125,50.4,1.3
22035,Toyota,Urban Cruiser,2012,4995,Manual,46000,Diesel,125,57.6,1.4
22036,Toyota,Urban Cruiser,2011,3995,Manual,60700,Petrol,125,50.4,1.3


In [35]:
# Step 2: sorting
Toyota_car.groupby('year').count()[['brand']].sort_values('brand', ascending=False).head(1).index[0]

np.int64(2017)

## **No.15**
Display the number of Toyota cars released in 2019 for each model

In [36]:
# table containing cars with the brand 'Toyota' manufactured in 2019
toyota_2019 = df_car[(df_car['year']==2019) & (df_car['brand']=='Toyota')]
toyota_2019

,brand,model,year,price,transmission,mileage,fuelType,tax,mpg,engineSize
15352,Toyota,GT86,2019,23995,Manual,995,Petrol,145,33.2,2.0
15354,Toyota,GT86,2019,23980,Manual,1751,Petrol,145,33.2,2.0
15357,Toyota,GT86,2019,23495,Automatic,3934,Petrol,145,32.8,2.0
15358,Toyota,GT86,2019,25780,Manual,5123,Petrol,145,33.2,2.0
15360,Toyota,GT86,2019,23998,Semi-Auto,913,Petrol,145,32.8,2.0
...,...,...,...,...,...,...,...,...,...,...
22018,Toyota,Camry,2019,26491,Automatic,4416,Hybrid,135,52.3,2.5
22019,Toyota,Camry,2019,27550,Automatic,4189,Hybrid,135,52.3,2.5
22020,Toyota,Camry,2019,25490,Automatic,6340,Hybrid,135,52.3,2.5
22021,Toyota,Camry,2019,26990,Automatic,3000,Hybrid,135,52.3,2.5


In [37]:
toyota_2019.groupby(['year', 'brand', 'model'], as_index=False)[['price']].count().rename(columns={'price':'amount'})

,year,brand,model,amount
0,2019,Toyota,Auris,3
1,2019,Toyota,Aygo,362
2,2019,Toyota,C-HR,108
3,2019,Toyota,Camry,10
4,2019,Toyota,Corolla,225
5,2019,Toyota,GT86,19
6,2019,Toyota,Hilux,19
7,2019,Toyota,Land Cruiser,15
8,2019,Toyota,PROACE VERSO,13
9,2019,Toyota,Prius,37


## **No. 16**

Find the most fuel-efficient Toyota model

In [38]:
# create a table containing only Toyota cars
Toyota_car = df_car[df_car['brand']=='Toyota']
Toyota_car

,brand,model,year,price,transmission,mileage,fuelType,tax,mpg,engineSize
15339,Toyota,GT86,2016,16000,Manual,24089,Petrol,265,36.2,2.0
15340,Toyota,GT86,2017,15995,Manual,18615,Petrol,145,36.2,2.0
15341,Toyota,GT86,2015,13998,Manual,27469,Petrol,265,36.2,2.0
15342,Toyota,GT86,2017,18998,Manual,14736,Petrol,150,36.2,2.0
15343,Toyota,GT86,2017,17498,Manual,36284,Petrol,145,36.2,2.0
...,...,...,...,...,...,...,...,...,...,...
22033,Toyota,IQ,2011,5500,Automatic,30000,Petrol,20,58.9,1.0
22034,Toyota,Urban Cruiser,2011,4985,Manual,36154,Petrol,125,50.4,1.3
22035,Toyota,Urban Cruiser,2012,4995,Manual,46000,Diesel,125,57.6,1.4
22036,Toyota,Urban Cruiser,2011,3995,Manual,60700,Petrol,125,50.4,1.3


In [39]:
# Find the maximum mpg (fuel-efficient car) for a Toyota car
max_mpg = Toyota_car['mpg'].max()
max_mpg

235.0

In [40]:
Toyota_car[Toyota_car['mpg']==max_mpg].iloc[0, 1]

'Prius'

## **No.17**

Show cars whose model names contain the letter 'x'!

In [41]:
# method 1
df_car[df_car['model'].str.contains('x|X')]['model'].unique()

array(['IX35', 'IX20', 'Hilux'], dtype=object)

In [43]:
# Method 2: Using apply

def check_x(name):
  if 'x' in name.lower():
    return True
  else:
    return False

In [44]:
df_car[df_car['model'].apply(check_x)]['model'].unique()

array(['IX35', 'IX20', 'Hilux'], dtype=object)

## **No.18**

Create a new dataframe with a pivot table containing the average tax for each transmission and fuel type.

In [45]:
df_car.pivot_table(
    index = 'transmission',
    columns = 'fuelType',
    values = 'tax',
    aggfunc = 'mean'
).fillna(0)

fuelType,Diesel,Hybrid,Other,Petrol
transmission,,,,
Automatic,153.920570,67.691480,104.405941,148.431250
Manual,88.033135,138.400000,30.000000,110.170476
Other,0.000000,0.000000,0.000000,145.000000
Semi-Auto,143.781060,122.192982,0.000000,142.927273


## **No.19**

Create a new dataframe with a pivot table containing the average price for each brand and transmission with a year column.

In [46]:
df_car.pivot_table(
    index = ['brand', 'transmission'],
    columns = 'year',
    values = 'price',
    aggfunc = 'mean'
).fillna(0).round(0)

year                    1997     1998    1999    2000   2001    2002    2003  \
brand   transmission                                                           
Audi    Automatic     4650.0   4999.0     0.0     0.0    0.0     0.0  3492.0   
        Manual           0.0      0.0     0.0     0.0    0.0  2433.0  3325.0   
        Semi-Auto        0.0      0.0     0.0     0.0    0.0     0.0     0.0   
Hyundai Automatic        0.0      0.0     0.0  1295.0    0.0     0.0  2192.0   
        Manual           0.0      0.0     0.0     0.0    0.0  1200.0  3290.0   
        Other            0.0      0.0     0.0     0.0    0.0     0.0     0.0   
        Semi-Auto        0.0      0.0     0.0     0.0    0.0     0.0     0.0   
Toyota  Automatic        0.0      0.0  1995.0  2695.0  899.0  2347.0  2295.0   
        Manual           0.0  19990.0     0.0     0.0    0.0  1275.0  1829.0   
        Other            0.0      0.0     0.0     0.0    0.0     0.0     0.0   
        Semi-Auto        0.0      0.0     0.0     0.0    0.0     0.0     0.0   

year                    2004    2005    2006  ...     2011     2012     2013  \
brand   transmission                          ...                              
Audi    Automatic     6891.0  5895.0  6459.0  ...  11486.0  12304.0  13398.0   
        Manual        2297.0  3090.0  6531.0  ...   7586.0   8692.0  10212.0   
        Semi-Auto        0.0     0.0     0.0  ...  12950.0  10959.0  14299.0   
Hyundai Automatic        0.0     0.0     0.0  ...   6396.0   7243.0   7729.0   
        Manual           0.0  2895.0  3490.0  ...   4984.0   4885.0   5483.0   
        Other            0.0     0.0     0.0  ...      0.0      0.0      0.0   
        Semi-Auto        0.0     0.0     0.0  ...      0.0      0.0  13161.0   
Toyota  Automatic     4782.0  2495.0  4995.0  ...   6642.0   8263.0  10934.0   
        Manual        1672.0  3105.0  2690.0  ...   4641.0   5820.0   6597.0   
        Other            0.0     0.0     0.0  ...      0.0      0.0      0.0   
        Semi-Auto        0.0     0.0     0.0  ...      0.0      0.0   8236.0   

year                     2014     2015     2016     2017     2018     2019  \
brand   transmission                                                         
Audi    Automatic     17549.0  18028.0  19717.0  24324.0  32357.0  33535.0   
        Manual        11544.0  12722.0  14301.0  16098.0  18583.0  23427.0   
        Semi-Auto     15758.0  18446.0  20567.0  23157.0  28997.0  31594.0   
Hyundai Automatic      8943.0  10272.0  13453.0  14474.0  16104.0  21768.0   
        Manual         6568.0   8112.0   9939.0  10460.0  11774.0  15265.0   
        Other             0.0      0.0      0.0  14745.0      0.0      0.0   
        Semi-Auto     12744.0  11914.0  13735.0  14919.0  16733.0  24440.0   
Toyota  Automatic     11732.0  12485.0  15065.0  16019.0  17140.0  22991.0   
        Manual         7038.0   7484.0   8691.0   9422.0  10063.0  12315.0   
        Other             0.0  12795.0      0.0      0.0      0.0      0.0   
        Semi-Auto      8915.0  13702.0  11956.0  12594.0  10004.0  25480.0   

year                     2020  
brand   transmission           
Audi    Automatic     39252.0  
        Manual        25096.0  
        Semi-Auto     38264.0  
Hyundai Automatic     24044.0  
        Manual        16152.0  
        Other             0.0  
        Semi-Auto     27759.0  
Toyota  Automatic     28744.0  
        Manual        15925.0  
        Other             0.0  
        Semi-Auto     21271.0  

[11 rows x 24 columns]

## **No.20**

How many models are there in this dataset with only one car?

In [47]:
sum(df_car['model'].value_counts()==1)

4